pkg/cmd/options/options.go

```diff
@@ -81,7 +84,7 @@ func (o *CustomMetricsAdapterServerOptions) AddFlags(fs *pflag.FlagSet) {
 }
 
 // ApplyTo applies CustomMetricsAdapterServerOptions to the server configuration.
-func (o *CustomMetricsAdapterServerOptions) ApplyTo(serverConfig *genericapiserver.Config) error {
+func (o *CustomMetricsAdapterServerOptions) ApplyTo(serverConfig *genericapiserver.Config, clientConfig *rest.Config, informers informers.SharedInformerFactory) error {
```

Comment:
Why not using a [RecommendedConfig](https://pkg.go.dev/k8s.io/apiserver/pkg/server#RecommendedConfig) here since it exactly, genericapiserver config + rest config + informer factory

Reply:
for some reason the client config is nil if we use this. Not only in the test but also when I build the container and run it in kind cluster.

Reply:
hmm, so that's what I was running into then I guess

Reply:
I'll try to spend some time investigating tomorrow, it's odd since that's what we are supposed to use

Reply:
Thanks, we would like to use this updated dep for our next release (scheduled for tomorrow 😄 )

@tomkerkhove FYI

Reply:
@zroubalik could you try applying this patch?

```diff
diff --git a/pkg/cmd/builder.go b/pkg/cmd/builder.go
index 1419378a..1571b148 100644
--- a/pkg/cmd/builder.go
+++ b/pkg/cmd/builder.go
@@ -303,13 +303,15 @@ func (b *AdapterBase) Config() (*apiserver.Config, error) {
                        return nil, err
                }
 
-               serverConfig := genericapiserver.NewConfig(apiserver.Codecs)
-               err = b.CustomMetricsAdapterServerOptions.ApplyTo(serverConfig, b.clientConfig, b.informers)
+               serverConfig := genericapiserver.NewRecommendedConfig(apiserver.Codecs)
+               serverConfig.ClientConfig = b.clientConfig
+               serverConfig.SharedInformerFactory = b.informers
+               err = b.CustomMetricsAdapterServerOptions.ApplyTo(serverConfig)
                if err != nil {
                        return nil, err
                }
                b.config = &apiserver.Config{
-                       GenericConfig: serverConfig,
+                       GenericConfig: &serverConfig.Config,
                }
        }
 
diff --git a/pkg/cmd/options/options.go b/pkg/cmd/options/options.go
index dcd340b0..12ff5a4e 100644
--- a/pkg/cmd/options/options.go
+++ b/pkg/cmd/options/options.go
@@ -25,9 +25,7 @@ import (
 
        genericapiserver "k8s.io/apiserver/pkg/server"
        genericoptions "k8s.io/apiserver/pkg/server/options"
-       "k8s.io/client-go/informers"
        "k8s.io/client-go/kubernetes"
-       "k8s.io/client-go/rest"
        openapicommon "k8s.io/kube-openapi/pkg/common"
 )
 
@@ -84,7 +82,7 @@ func (o *CustomMetricsAdapterServerOptions) AddFlags(fs *pflag.FlagSet) {
 }
 
 // ApplyTo applies CustomMetricsAdapterServerOptions to the server configuration.
-func (o *CustomMetricsAdapterServerOptions) ApplyTo(serverConfig *genericapiserver.Config, clientConfig *rest.Config, informers informers.SharedInformerFactory) error {
+func (o *CustomMetricsAdapterServerOptions) ApplyTo(serverConfig *genericapiserver.RecommendedConfig) error {
        // TODO have a "real" external address (have an AdvertiseAddress?)
        if err := o.SecureServing.MaybeDefaultWithSelfSignedCerts("localhost", nil, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
                return fmt.Errorf("error creating self-signed certificates: %v", err)
@@ -99,16 +97,16 @@ func (o *CustomMetricsAdapterServerOptions) ApplyTo(serverConfig *genericapiserv
        if err := o.Authorization.ApplyTo(&serverConfig.Authorization); err != nil {
                return err
        }
-       if err := o.Audit.ApplyTo(serverConfig); err != nil {
+       if err := o.Audit.ApplyTo(&serverConfig.Config); err != nil {
                return err
        }
 
-       clientset, err := kubernetes.NewForConfig(clientConfig)
+       clientset, err := kubernetes.NewForConfig(serverConfig.ClientConfig)
        if err != nil {
                return err
        }
 
-       if err := o.Features.ApplyTo(serverConfig, clientset, informers); err != nil {
+       if err := o.Features.ApplyTo(&serverConfig.Config, clientset, serverConfig.SharedInformerFactory); err != nil {
                return err
        }
 
diff --git a/pkg/cmd/options/options_test.go b/pkg/cmd/options/options_test.go
index 862c904f..c5cc16db 100644
--- a/pkg/cmd/options/options_test.go
+++ b/pkg/cmd/options/options_test.go
@@ -113,10 +113,11 @@ func TestApplyTo(t *testing.T) {
                        err := flagSet.Parse(c.args)
                        assert.NoErrorf(t, err, "Error while parsing flags")
 
-                       serverConfig := genericapiserver.NewConfig(apiserver.Codecs)
-                       clientConfig := &rest.Config{}
-                       informers := informers.NewSharedInformerFactory(nil, 0)
-                       err = o.ApplyTo(serverConfig, clientConfig, informers)
+                       serverConfig := genericapiserver.NewRecommendedConfig(apiserver.Codecs)
+                       serverConfig.ClientConfig = &rest.Config{}
+                       serverConfig.SharedInformerFactory = informers.NewSharedInformerFactory(nil, 0)
+
+                       err = o.ApplyTo(serverConfig)
 
                        defer func() {
                                // Close the listener, if any
```

at least the unit tests are passing, I'll try with prometheus-adapter e2e tests

Reply:
@dgrisonnet on it

Reply:
I ran into the same rbac failures locally with prometheus-adapter. It seems like featuregates need additional rbac permissions, let me check what we should do about that

Reply:
https://github.com/kubernetes/kubernetes/commit/b3216214f95302aa886697eb72349b5b2f5cbe5a

Adding perms to flowschema should do it

Reply:
Yeah, found the same problem.

Reply:
Fixed the testing adapter, it was failing there as well: c46dc0631a852156be55131a6ae6be1710755e70

Reply:
I was also on an outdated Kind cluster, so I am trying again with 1.29

Reply:
Also in KEDA, we are good to merge @dgrisonnet